app/src/main/java/com/hedvig/app/feature/checkout/EditQuotesUseCase.kt

```diff
@@ -0,0 +1,123 @@
+import org.json.JSONObject
+
+class EditQuotesUseCase(
+    private val apolloClient: ApolloClient,
+    private val featureManager: FeatureManager,
+    private val localeManager: LocaleManager,
+    private val graphQLQueryHandler: GraphQLQueryHandler,
+) {
+
+    data class Error(val message: String? = null)
+
+    object Success
+
+    suspend fun editQuotes(parameter: EditAndSignParameter): Either<Error, Success> = parameter.quoteIds
+        .map { editQuote(it, parameter.quoteCartId, parameter.ssn, parameter.email) }
+        .sequenceEither()
+        .map { Success }
+
+    private suspend fun editQuote(
+        quoteId: String,
+        quoteCartId: String?,
+        ssn: String,
+        email: String
+    ): Either<Error, Success> {
+        return if (featureManager.isFeatureEnabled(Feature.QUOTE_CART)) {
+            if (quoteCartId == null) {
+                Either.Left(Error("No quote cart id found"))
+            } else {
+                editQuoteCart(quoteCartId, quoteId, ssn, email)
+            }
+        } else {
+            editQuoteWithEmailAndSSN(quoteId, ssn, email)
+        }
+    }
+
+    private suspend fun editQuoteCart(
+        quoteCartId: String,
+        quoteId: String,
+        ssn: String,
+        email: String
+    ): Either<Error, Success> {
+        val json = buildJsonObject {
+            put("quoteCartId", quoteCartId)
+            put("quoteId", quoteId)
+            put("locale", localeManager.defaultLocale().rawValue)
+            putJsonObject("payload") {
+                put("ssn", ssn)
+                put("email", email)
+                putJsonObject("data") {
+                    put("ssn", ssn)
+                    put("email", email)
+                }
+            }
+        }.toString()
+
+        return graphQLQueryHandler.graphQLQuery(
+            query = QuoteCartEditQuoteMutation.QUERY_DOCUMENT,
+            variables = JSONObject(json),
+            files = emptyList()
+        )
+            .toEither()
+            .mapLeft { Error() }
+            .flatMap { jsonResponse ->
+                val errorCode = try {
+                    jsonResponse.getJSONObject("data")
+                        .getJSONObject("quoteCart_editQuote")
+                        .getJSONArray("limits")
+                        .getJSONObject(0)
+                        .getString("code")
+                } catch (exception: JSONException) {
+                    null
+                }
+
+                if (errorCode != null) {
+                    Either.Left(Error(errorCode))
+                } else {
+                    Either.Right(Success)
+                }
+            }
+    }
+
+    private suspend fun editQuoteWithEmailAndSSN(
+        quoteId: String,
+        ssn: String,
+        email: String
+    ): Either<Error, Success> = apolloClient.mutate(EditMailAndSSNMutation(quoteId, ssn, email))
+        .safeQuery()
+        .toEither()
+        .mapLeft { Error(it.message) }
```

Comment:
We have a `toEither` version that looks like this:
```kotlin
inline fun <ErrorType> toEither(ifEmpty: (message: String?) -> ErrorType): Either<ErrorType, T> = when (this) {
    is Error -> Either.Left(ifEmpty(message))
    is Success -> Either.Right(this.data)
}
```
which could potentially make this instead be 
```kotlin
.toEither { Error(it) }
```

This is to replicate the behavior that exists in the arrow library on the `Option` type:
```kotlin
public inline fun <L> toEither(ifEmpty: () -> L): Either<L, A>
```
Where they also have a toEither where you can in the same function do the mapping of what your error type would be, without having to do a `mapLeft` afterwards. Plus in our case you get access to message directly as you may want to use it in the new error type (like you do in this case with `it.message`)

Reply:
Awesome